File: src/5v2.py
import random
from sys import stdin, stdout
from collections import deque
import itertools

def readln():
    return stdin.readline().rstrip()

def outln(n):
    stdout.write(str(n) + '\n')

# Direcoes possiveis para se mover no labirinto: cima, baixo, esquerda, direita
DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]

def bfs_path(maze, start, goal, water_filled):
    n, m = len(maze), len(maze[0])
    queue = deque([start])
    came_from = {start: None}

    while queue:
        current = queue.popleft()
        if current == goal:
            return reconstruct_path(came_from, current)

        for direction in DIRECTIONS:
            neighbor = (current[0] + direction[0], current[1] + direction[1])
            if 0 <= neighbor[0] < n and 0 <= neighbor[1] < m and maze[neighbor[0]][neighbor[1]] != '#' and not water_filled[neighbor[0]][neighbor[1]]:
                if neighbor not in came_from:
                    queue.append(neighbor)
                    came_from[neighbor] = current
    return None

def reconstruct_path(came_from, current):
    total_path = [current]
    while came_from[current] is not None:
        current = came_from[current]
        total_path.append(current)
    total_path.reverse()
    return total_path

def flood_simulation(grid, start_points, block=None):
    rows, columns = len(grid), len(grid[0])
    visited = [[False] * columns for _ in range(rows)]
    queue = start_points[:]
    
    for x, y in queue:
        visited[x][y] = True

    i = 0
    while i < len(queue):
        x, y = queue[i]
        i += 1

        for dx, dy in DIRECTIONS:
            nx, ny = x + dx, y + dy
            if grid[nx][ny] == '#' or visited[nx][ny]:
                continue

            if 0 <= nx < rows and 0 <= ny < columns and grid[nx][ny] == '.' and not visited[nx][ny]:
                if block and (((x, y) == block[0] and (nx, ny) == block[1]) or ((x, y) == block[1] and (nx, ny) == block[0])):
                    continue

                visited[nx][ny] = True
                queue.append((nx, ny))

    return visited



def dfs_bridges(x, y, grid, visited, disc, low, parent, bridges, time):
    visited[x][y] = True
    disc[x][y] = low[x][y] = time[0]
    time[0] += 1
    
    for dx, dy in DIRECTIONS:
        nx, ny = x + dx, y + dy
        if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and grid[nx][ny] == '.':
            if not visited[nx][ny]:
                parent[nx][ny] = (x, y)
                dfs_bridges(nx, ny, grid, visited, disc, low, parent, bridges, time)
                
                low[x][y] = min(low[x][y], low[nx][ny])
                
                if low[nx][ny] > disc[x][y]:
                    bridges.append(((x, y), (nx, ny)))
            elif (nx, ny) != parent[x][y]:
                low[x][y] = min(low[x][y], disc[nx][ny])


def find_bridges(grid):
    rows, columns = len(grid), len(grid[0])
    visited = [[False] * columns for _ in range(rows)]
    disc = [[float('inf')] * columns for _ in range(rows)]
    low = [[float('inf')] * columns for _ in range(rows)]
    parent = [[None] * columns for _ in range(rows)]
    bridges = []
    time = [0]
    
    for i in range(rows):
        for j in range(columns):
            if (grid[i][j] == 'M' or grid[i][j] == '.') and not visited[i][j]:
                dfs_bridges(i, j, grid, visited, disc, low, parent, bridges, time)
    
    return bridges

def find_dominating_set_for_flood_control(grid, manholes, bridges):
    rows, columns = len(grid), len(grid[0])
    initial_flooded_areas = flood_simulation(grid, manholes)
    
    bridge_best = (-1, -1)
    count_inicial = sum([1 for i in range(len(initial_flooded_areas)) for j in range(len(initial_flooded_areas[0])) if initial_flooded_areas[i][j] == True])
    for bridge in bridges:
        blocked_flooded_areas = flood_simulation(grid, manholes, block=bridge)
        count = sum([1 for i in range(len(blocked_flooded_areas)) for j in range(len(blocked_flooded_areas[0])) if blocked_flooded_areas[i][j] == True])

        if count < count_inicial:
            count_inicial = count
            bridge_best = bridge

    return bridge_best


def main():
    num_cases = int(readln())
    results = []

    for _ in range(num_cases):
        n, m = map(int, readln().split())
        maze = [list(readln()) for _ in range(n)]
        num_covers = int(readln())

        door = None
        exit = None
        manholes = []

        for r in range(n):
            for c in range(m):
                if maze[r][c] == 'D':
                    door = (r, c)
                elif maze[r][c] == 'E':
                    exit = (r, c)
                elif maze[r][c] == 'M':
                    manholes.append((r, c))
        
        bridges = find_bridges(maze)
        
        solution = []
        for cover_combination in itertools.combinations(manholes, num_covers):
            manhole_aux = list(set(manholes) - set(cover_combination))
            dominating_set = find_dominating_set_for_flood_control(maze, manhole_aux, bridges)
            water_filled = flood_simulation(maze, manhole_aux, dominating_set)
                
            path = bfs_path(maze, door, exit, water_filled)
            if path:
                solution.append([dominating_set, cover_combination, path])
            else:
                solution.append([dominating_set, cover_combination, []])

        flood_gate, covers, path = random.choice(solution)
        outln(f"{flood_gate[0][0]} {flood_gate[0][1]} {flood_gate[1][0]} {flood_gate[1][1]}")
        outln(str(len(covers)))
        for cover in covers:
            outln(f"{cover[0]} {cover[1]}")
        outln(str(len(path)))
        for step in path:
            outln(f"{step[0]} {step[1]}")

if __name__ == '__main__':
    main()

File: src/5v2.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <utility>
#include <tuple>
#include <set>
#include <map>
#include <sstream>
#include <ctime>
#include <climits>

using namespace std;

const vector<pair<int, int>> DIRECTIONS = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

vector<string> read_maze(int N, int M) {
    vector<string> maze(N);
    for (int i = 0; i < N; ++i) {
        cin >> maze[i];
    }
    return maze;
}

bool valid_flood_gate(vector<string>& maze, int N, int M, int sx, int sy, int ex, int ey, pair<pair<int, int>, pair<int, int>> bridge);

vector<pair<pair<int, int>, pair<int, int>>> dfs_bridges_iterative(vector<string>& grid, int sx, int sy, int ex, int ey) {
    int rows = grid.size();
    int columns = grid[0].size();
    vector<vector<bool>> visited(rows, vector<bool>(columns, false));
    vector<vector<int>> disc(rows, vector<int>(columns, INT_MAX));
    vector<vector<int>> low(rows, vector<int>(columns, INT_MAX));
    vector<vector<pair<int, int>>> parent(rows, vector<pair<int, int>>(columns, {-1, -1}));
    vector<pair<pair<int, int>, pair<int, int>>> bridges;
    stack<tuple<int, int, int>> st;
    int time = 0;

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < columns; ++j) {
            if ((grid[i][j] == 'M' || grid[i][j] == '.') && !visited[i][j]) {
                st.push({i, j, 0});

                while (!st.empty()) {
                    auto [x, y, state] = st.top();
                    st.pop();

                    if (state == 0) {
                        visited[x][y] = true;
                        disc[x][y] = low[x][y] = time++;
                        st.push({x, y, 1});

                        for (auto [dx, dy] : DIRECTIONS) {
                            int nx = x + dx, ny = y + dy;
                            if (nx >= 0 && nx < rows && ny >= 0 && ny < columns && (grid[nx][ny] == '.' || grid[nx][ny] == 'M')) {
                                if (!visited[nx][ny]) {
                                    st.push({nx, ny, 0});
                                    parent[nx][ny] = {x, y};
                                } else if (make_pair(nx, ny) != parent[x][y]) {
                                    low[x][y] = min(low[x][y], disc[nx][ny]);
                                }
                            }
                        }
                    } else {
                        for (auto [dx, dy] : DIRECTIONS) {
                            int nx = x + dx, ny = y + dy;
                            if (nx >= 0 && nx < rows && ny >= 0 && ny < columns && (grid[nx][ny] == '.' || grid[nx][ny] == 'M')) {
                                if (parent[nx][ny] == make_pair(x, y)) {
                                    low[x][y] = min(low[x][y], low[nx][ny]);
                                    if (low[nx][ny] > disc[x][y]) {
                                        pair<int, int> p1 = {x, y};
                                        pair<int, int> p2 = {nx, ny};
                                        if (p1 < p2 && valid_flood_gate(grid, rows, columns, sx, sy, ex, ey, {p1, p2})) {
                                            bridges.push_back({p1, p2});
                                        } else if (p1 >= p2 && valid_flood_gate(grid, rows, columns, sx, sy, ex, ey, {p2, p1})) {
                                            bridges.push_back({p2, p1});
                                        }
                                    }
                                } else if (make_pair(nx, ny) != parent[x][y]) {
                                    low[x][y] = min(low[x][y], disc[nx][ny]);
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    return bridges;
}

vector<pair<int, int>> bfs_safe_path(vector<string>& maze, int N, int M, int sx, int sy, int ex, int ey) {
    queue<pair<int, int>> q;
    vector<vector<bool>> visited(N, vector<bool>(M, false));
    vector<vector<pair<int, int>>> parent(N, vector<pair<int, int>>(M, {-1, -1}));
    q.push({sx, sy});
    visited[sx][sy] = true;

    while (!q.empty()) {
        auto [x, y] = q.front();
        q.pop();
        if (x == ex && y == ey) {
            vector<pair<int, int>> path;
            while (parent[x][y] != make_pair(-1, -1)) {
                path.push_back({x, y});
                tie(x, y) = parent[x][y];
            }
            path.push_back({sx, sy});
            reverse(path.begin(), path.end());
            return path;
        }
        for (auto [dx, dy] : DIRECTIONS) {
            int nx = x + dx, ny = y + dy;
            if (nx >= 0 && nx < N && ny >= 0 && ny < M && !visited[nx][ny] && maze[nx][ny] != '#') {
                visited[nx][ny] = true;
                parent[nx][ny] = {x, y};
                q.push({nx, ny});
            }
        }
    }
    return {};
}


unordered_map<tuple<set<pair<int, int>>, pair<pair<int, int>, pair<int, int>>>, vector<vector<bool>>> memo(1);

vector<vector<bool>> bfs_flood_control(vector<string>& grid, const set<pair<int, int>>& start_points, pair<pair<int, int>, pair<int, int>> flood_gate, const set<pair<int, int>>& block) {
    auto key = make_tuple(start_points, flood_gate, block);
    if (memo.find(key) != memo.end()) {
        return memo[key];
    }

    int rows = grid.size();
    int columns = grid[0].size();
    queue<pair<int, int>> q;
    vector<vector<bool>> visited(rows, vector<bool>(columns, false));

    for (auto& sp : start_points) {
        q.push(sp);
    }

    for (auto& b : block) {
        if (q.front() == b) {
            q.pop();
        }
    }

    while (!q.empty()) {
        auto [x, y] = q.front();
        q.pop();

        for (auto [dx, dy] : DIRECTIONS) {
            int nx = x + dx, ny = y + dy;
            if (nx >= 0 && nx < rows && ny >= 0 && ny < columns && !visited[nx][ny] && grid[nx][ny] != '#') {
                if (block.count({nx, ny}) && ((dx == 0 && (dy == -1 || dy == 1)) || (dy == 0 && (dx == -1 || dx == 1))) &&
                    make_pair(nx, ny) == flood_gate.first && make_pair(x, y) == flood_gate.second) {
                    continue;
                }
                visited[nx][ny] = true;
                q.push({nx, ny});
            }
        }
    }

    memo[key] = visited;
    return visited;
}

unordered_map<vector<pair<int, int>>, int> evaluate_manhole_cover_effectiveness(vector<string>& maze, int N, int M, vector<pair<int, int>>& manholes, int sx, int sy, pair<pair<int, int>, pair<int, int>> flood_gate, int C) {
    unordered_map<vector<pair<int, int>>, int> flood_effectiveness;
    set<pair<int, int>> manhole_set(manholes.begin(), manholes.end());

    auto base_flooded = bfs_flood_control(maze, manhole_set, flood_gate, {});

    auto combination = [](vector<pair<int, int>>& elements, int K) {
        vector<vector<pair<int, int>>> result;
        vector<bool> v(elements.size());
        fill(v.end() - K, v.end(), true);
        do {
            vector<pair<int, int>> combination;
            for (size_t i = 0; i < elements.size(); ++i) {
                if (v[i]) {
                    combination.push_back(elements[i]);
                }
            }
            result.push_back(combination);
        } while (next_permutation(v.begin(), v.end()));
        return result;
    };

    for (auto& comb : combination(manholes, C)) {
        set<pair<int, int>> comb_set(comb.begin(), comb.end());
        auto flood = bfs_flood_control(maze, manhole_set, flood_gate, comb_set);

        vector<pair<int, int>> filtered_comb;
        for (auto& m : comb) {
            if (!flood[m.first][m.second]) {
                filtered_comb.push_back(m);
            }
        }

        flood_effectiveness[filtered_comb] = 0;
        for (int x = 0; x < N; ++x) {
            for (int y = 0; y < M; ++y) {
                if (base_flooded[x][y] && !flood[x][y]) {
                    flood_effectiveness[filtered_comb]++;
                }
            }
        }
    }

    return flood_effectiveness;
}

vector<pair<int, int>> select_effective_covers(unordered_map<vector<pair<int, int>>, int>& manhole_effectiveness) {
    auto best_combination = max_element(manhole_effectiveness.begin(), manhole_effectiveness.end(),
                                        [](const auto& a, const auto& b) { return a.second < b.second; })->first;
    return best_combination;
}

bool valid_flood_gate(vector<string>& maze, int N, int M, int sx, int sy, int ex, int ey, pair<pair<int, int>, pair<int, int>> bridge) {
    auto [p1, p2] = bridge;
    maze[p1.first][p1.second] = '#';
    maze[p2.first][p2.second] = '#';

    bool path_exists = !bfs_safe_path(maze, N, M, sx, sy, ex, ey).empty();

    maze[p1.first][p1.second] = '.';
    maze[p2.first][p2.second] = '.';

    return path_exists;
}

int main() {
    int num_cases;
    cin >> num_cases;
    while (num_cases--) {
        int N, M;
        cin >> N >> M;
        auto maze = read_maze(N, M);
        int C;
        cin >> C;

        pair<int, int> ex, ey;
        pair<int, int> dx, dy;
        vector<pair<int, int>> manholes;

        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                if (maze[i][j] == 'E') {
                    ex = {i, j};
                } else if (maze[i][j] == 'D') {
                    dx = {i, j};
                } else if (maze[i][j] == 'M') {
                    manholes.push_back({i, j});
                }
            }
        }

        auto bridges = dfs_bridges_iterative(maze, dx.first, dx.second, ex.first, ex.second);
        pair<pair<int, int>, pair<int, int>> flood_gate = bridges[rand() % bridges.size()];

        auto manhole_effectiveness = evaluate_manhole_cover_effectiveness(maze, N, M, manholes, dx.first, dx.second, flood_gate, C);
        auto selected_covers = select_effective_covers(manhole_effectiveness);

        cout << flood_gate.first.first << " " << flood_gate.first.second << " " << flood_gate.second.first << " " << flood_gate.second.second << endl;
        cout << selected_covers.size() << endl;
        for (auto& cover : selected_covers) {
            cout << cover.first << " " << cover.second << endl;
        }
        auto path = bfs_safe_path(maze, N, M, dx.first, dx.second, ex.first, ex.second);
        cout << path.size() << endl;
        for (auto& step : path) {
            cout << step.first << " " << step.second << endl;
        }
    }
    return 0;
}
